### PR TITLE
- add tools for Apple development

### LINKED ISSUE #2322: Suggestion: adding a helper for darwin host
Hi,

I was wondering if a helper could be added in `tools` to detect if the `os` setting is targeting a darwin platform. Here is my use-case:

```python
    if self.settings.os in ["Macos", "iOS", "watchOS",  "tvOS"]:
        # ...

    # What I have in mind
    if tools.darwin_platform(self.settings.os):
        # ...
```

This is not a big deal, but I'm too lazy to type it in every recipe :D

### PR DESCRIPTION
**Pull request by SSE4**
add various tools I am using for apple development
I was using them for my private recipes and keep copy-pasting between packages
now I am compiling bincrafters packages for iOS, and need them as well
so I've decided it is beneficial to have tools in conan codebase

Closes #2322

### CODE REVIEW

File: conans/client/tools/apple.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+import subprocess
+
+
+def cmd_output(command):
```
Comment (lasote) [inline conans/client/tools/apple.py]: it doesn't look like an apple thing

File: conans/client/tools/apple.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    """converts conan-style architecture into Apple-style arch"""
+    return {'x86': 'i386',
+            'x86_64': 'x86_64',
+            'armv7': 'armv7',
+            'armv8': 'arm64',
+            'armv7s': 'armv7s',
+            'armv7k': 'armv7k'}.get(str(arch))
+
+
+def apple_sdk_name(settings):
```
Comment (lasote) [inline conans/client/tools/apple.py]: sdk name for what? CMake?

File: conans/client/tools/apple.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+                'tvOS': 'appletvsimulator'}.get(str(_os))
+    elif str(arch).startswith('arm'):
+        return {'iOS': 'iphoneos',
+                'watchOS': 'watchos',
+                'tvOS': 'appletvos'}.get(str(_os))
+    else:
+        return None
+
+
+def apple_deployment_target_env_name(_os):
```
Comment (lasote) [inline conans/client/tools/apple.py]: What are these variables used for? CMake related? XCode related?...

File: conans/client/tools/apple.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    """environment variable name which controls deployment target"""
+    return {'Macos': 'MACOSX_DEPLOYMENT_TARGET',
+            'iOS': 'IOS_DEPLOYMENT_TARGET',
+            'watchOS': 'WATCHOS_DEPLOYMENT_TARGET',
+            'tvOS': 'TVOS_DEPLOYMENT_TARGET'}.get(str(_os))
+
+
+def apple_deployment_target_flag_name(_os):
```
Comment (lasote) [inline conans/client/tools/apple.py]: Same as above

File: conans/client/tools/apple.py (hunk 1 above)
Comment (SSE4) [reply conans/client/tools/apple.py]: where to put it?

File: conans/client/tools/apple.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    """compiler flag name which controls deployment target"""
+    return {'Macos': '-mmacosx-version-min',
+            'iOS': '-mios-version-min',
+            'watchOS': '-mwatchos-version-min',
+            'tvOS': '-mappletvos-version-min'}.get(str(_os))
+
+
+class XCRun:
```
Comment (lasote) [inline conans/client/tools/apple.py]: Please, elaborate a use case.

File: conans/client/tools/apple.py (hunk 2 above)
Comment (SSE4) [reply conans/client/tools/apple.py]: for xcrun (-sdk option)
rename to something like apple_xcrun_sdk_name?

File: conans/client/tools/apple.py (hunk 4 above)
Comment (SSE4) [reply conans/client/tools/apple.py]: it's compiler flag used by g++/clang++

**Comment by lasote** [comment]
Sorry because I do not know almost anything about iOS and Apple stuff, but without context, the new apple module looks like a drop box. 
Some tests are needed too.

File: conans/client/tools/apple.py (hunk 3 above)
Comment (SSE4) [reply conans/client/tools/apple.py]: in certain projects (e.g. log4cplus, ffmpeg) ./configure use them

File: conans/client/tools/apple.py (hunk 5 above)
Comment (SSE4) [reply conans/client/tools/apple.py]: for instance, cross-comping to iOS requires `-isysroot` option to be passed
so you could use it like:
```
sdk_name = apple_sdk_name(self.settings) # e.g. iphoneos
xcrun = XCRun(sdk_name)
CFLAGS = '--isysroot="%s"' % xcrun. sdk_path
```

**Comment by SSE4** [comment]
k, I'll add some tests

File: conans/client/tools/apple.py (hunk 1 above)
Comment (lasote) [reply conans/client/tools/apple.py]: Not sure, maybe we just need to improve the `conanfile.run` method and not add another one.

**Comment by SSE4** [comment]
I need some suggestion on how to design tests.
as most of code just executes `xcrun` and uses its output, how shall I test it?
shall I mock/fake xcrun or its output? or shall I directly invoke it and make tests OSX-only?

**Comment by memsharded** [comment]
Hi!

We need both. You can test the helper functions as unit-tests, as they are simple logic methods, no real action, I think no need to fake/mock

Then, at least one real test. We usually protect them with some ``if platform.system() != "Darwin": return``, so you can do that too (we will improve that with something more elegant at some point)

File: conans/client/tools/apple.py (hunk 5 above)
Comment (memsharded) [inline conans/client/tools/apple.py]: Modern python recommends ``XCRun(object)``

Comment (SSE4) [reply conans/client/tools/apple.py]: done

**Comment by lasote** [comment]
Sorry I had to cancel the builds, we are trying to get ready the release and having some problems and big delays, please, write me tomorrow if I forget to relaunch them.
Thanks so much and sorry for that.

**Comment by SSE4** [comment]
@lasote sure, no problem!

**Comment by SSE4** [comment]
@memsharded rebase, pls review

**Comment by memsharded** [comment]
I think we should move this forward. @theodelrieu and others are interested in these tools. We will need help to mature, and probably want to introduce them as "experimental", but lets aim to include them at 1.2

**Comment by SSE4** [comment]
@memsharded good news, let's revisit it
lemme know if something has to be changed according to new conan 1.0 features

**Comment by GeorgeScrivener** [comment]
@SSE4 do you have an example of how to use this?

**Comment by SSE4** [comment]
@GeorgeScrivener I have some from my private recipes
OpenSSL:
```
                sdk = tools.apple_sdk_name(self.settings)
                xcrun = tools.XCRun(sdk)
                env_vars["CROSS_SDK"] = os.path.basename(xcrun.sysroot)
                env_vars["CROSS_TOP"] = os.path.dirname(os.path.dirname(xcrun.sysroot))
                env_vars["CC"] = xcrun.cc
```
libxml2:
```
            sdk = tools.apple_sdk_name(self.settings)
            xcrun = tools.XCRun(sdk)
            apple_arch = tools.to_apple_arch(self.settings.arch)
            env_build.flags.append('-fembed-bitcode')
            env_build.flags.append('-isysroot %s' % xcrun.sysroot)
            env_build.flags.append('-arch %s' % apple_arch)
```
others are similar

**Comment by lasote** [comment]
@SSE4 Sorry for the long delay of this PR, I want to incorporate these features in the 1.4. 
I have some doubts, not about your code and its usage, but where to put the tools, maybe more integrated with the build helpers etc. So I will come back with a PR for you or questions or suggestions to finally merge this ;)
Thanks, as always a great contribution.

**Comment by lasote** [comment]
@SSE4 Could you update this PR from develop branch? I would like to open a PR for you and see what you think.

**Comment by SSE4** [comment]
@lasote sure, will do on evening
P.S. I've asked bincrafters representative (@solvingj) to discuss this PR on SwarmUP conference with conan developer in person, and conference happens in May 16 if I am not mistaken.

**Comment by lasote** [comment]
Almost the same question for `apple_deployment_target_env_var`, the value of the env vars have the same meaning than `apple_deployment_target_flag_name`?

**Comment by SSE4** [comment]
right

**Comment by lasote** [comment]
In principle `apple_sdk_name` is only for the XCRun or you have used isolated so far?

**Comment by SSE4** [comment]
I don't remember any isolated usage, probably can be safely hidden into XCRun

**Comment by SSE4** [comment]
ignore my previous comment - I have found the valid usage - for CMAKE_XCODE_EFFECTIVE_PLATFORMS during the CMake toolchain file generation.

**Comment by SSE4** [comment]
did fresh rebase on top of develop

**Comment by lasote** [comment]
@memsharded @danimtb everything looks good to me. Please review

File: conans/client/tools/apple.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+
+
+def apple_deployment_target_flag(os_, os_version):
+    """compiler flag name which controls deployment target"""
+    flag = {'Macos': '-mmacosx-version-min',
+            'iOS': '-mios-version-min',
+            'watchOS': '-mwatchos-version-min',
+            'tvOS': '-mappletvos-version-min'}.get(str(os_))
+    if not flag:
+        return None
```
[the remark this answers is not in the document]
Comment (SSE4) [reply conans/client/tools/apple.py]: I've changed to return an empty string by default

**Pull request merged**